File: server/src/main/java/neuroevolution/math/Vector.java
package neuroevolution.math;

import neuroevolution.exception.InvalidDimensionsException;

public class Vector {

	private final double[] elements;

	public Vector(final double... elements) {
		this.elements = elements;
	}

	public Vector(final int size, final VectorValueProvider valueProvider) {
		elements = new double[size];
		for (int i = 0; i < elements.length; ++i) {
			elements[i] = valueProvider.generateValue(i);
		}
	}

	public int length() {
		return elements.length;
	}

	public double get(final int index) {
		if (index < 0 || index >= elements.length) {
			throw new IndexOutOfBoundsException("Cannot access vector element. Index out of bound.");
		}

		return elements[index];
	}

	public Vector add(final Vector other) {
		if (elements.length != other.elements.length) {
			throw new InvalidDimensionsException("Cannot add vectors. Dimensions does not match.");
		}

		return new Vector(elements.length, index -> elements[index] + other.elements[index]);
	}

	public Vector divide(final double factor) {
		return new Vector(elements.length, index -> elements[index] / factor);
	}

	public Vector apply(final ValueModifier modifier) {
		return new Vector(elements.length, index -> modifier.apply(elements[index]));
	}

}

File: webapp/src/app/evolution/logic/neural-network/vector/vector.ts
export type VectorElementProvider = (index: number) => number;
export type VectorElementModifier = (element: number) => number;

export class Vector {

  get length(): number {
    return this.elements.length;
  }

  private constructor(public readonly elements: readonly number[]) {
  }

  static fromProvider(length: number, provider: VectorElementProvider): Vector {
    const elements: number[] = [];
    for (let i = 0; i < length; ++i) {
      const element: number = provider(i);
      if (!element && element !== 0) {
        throw new Error('Vector element cannot be empty.');
      }
      elements.push(element);
    }
    return new Vector(elements);
  }

  static fromArray(elements: readonly number[]): Vector {
    return new Vector(elements);
  }

  apply(modifier: VectorElementModifier): Vector {
    return Vector.fromProvider(
      this.length,
      (index: number): number => modifier(this.elements[index])
    );
  }

  add(vector: Vector): Vector {
    if (this.length !== vector.length) {
      throw new Error('Cannot add vectors. Lengths does not match.');
    }

    return Vector.fromProvider(
      this.length,
      (index: number): number => this.elements[index] + vector.elements[index]
    );
  }

}
